pkg/controllers/csi/provisioner/install.go

```diff
@@ -3,118 +3,119 @@ package csiprovisioner
-	urlInstaller := provisioner.urlInstallerBuilder(provisioner.fs, dtc, getUrlProperties(targetVersion, provisioner.path))
-
-	targetDir := provisioner.path.AgentSharedBinaryDirForAgent(targetVersion)
-	targetConfigDir := provisioner.path.AgentConfigDir(dk.GetName())
+	targetDir := provisioner.getTargetDir(dk)
 
-	err := provisioner.installAgent(ctx, urlInstaller, dk, targetDir, targetVersion)
+	_, err = agentInstaller.InstallAgent(ctx, targetDir)
 	if err != nil {
-		return "", err
+		return err
 	}
 
-	err = processmoduleconfig.CreateAgentConfigDir(provisioner.fs, targetConfigDir, targetDir, latestProcessModuleConfig)
+	err = provisioner.createLatestVersionSymlink(dk, targetDir)
 	if err != nil {
-		return "", err
+		return err
 	}
 
-	return targetVersion, nil
+	return provisioner.setupAgentConfigDir(ctx, dk, targetDir)
 }
 
-func (provisioner *OneAgentProvisioner) installAgent(ctx context.Context, agentInstaller installer.Installer, dk dynakube.DynaKube, targetDir, targetVersion string) error {
-	eventRecorder := updaterEventRecorder{
-		recorder: provisioner.recorder,
-		dk:       &dk,
+func (provisioner *OneAgentProvisioner) getInstaller(ctx context.Context, dk dynakube.DynaKube) (installer.Installer, error) {
+	switch {
+	case dk.CodeModulesVersion() != "":
+		dtc, err := buildDtc(provisioner, ctx, dk)
+		if err != nil {
+			return nil, err
+		}
+
+		props := &url.Properties{
+			Os:            dtclient.OsUnix,
+			Type:          dtclient.InstallerTypePaaS,
+			Arch:          arch.Arch,
+			Flavor:        arch.Flavor,
+			Technologies:  []string{"all"},
+			TargetVersion: dk.CodeModulesVersion(),
+			SkipMetadata:  true,
+			PathResolver:  provisioner.path,
+		}
+
+		urlInstaller := provisioner.urlInstallerBuilder(provisioner.fs, dtc, props)
+
+		return urlInstaller, nil
+	case dk.CodeModulesImage() != "":
+		props := &image.Properties{
+			ImageUri:     dk.CodeModulesImage(),
+			ApiReader:    provisioner.apiReader,
+			Dynakube:     &dk,
+			PathResolver: provisioner.path,
+		}
+
+		imageInstaller, err := provisioner.imageInstallerBuilder(ctx, provisioner.fs, props)
+		if err != nil {
+			return nil, err
+		}
+
+		return imageInstaller, nil
+	default:
+		return nil, errors.New("not enough information to download CodeModule")
 	}
-	isNewlyInstalled, err := agentInstaller.InstallAgent(ctx, targetDir)
-
-	if err != nil {
-		eventRecorder.sendFailedInstallAgentVersionEvent(targetVersion, dk.GetName())
+}
 
-		return err
+func (provisioner *OneAgentProvisioner) getTargetDir(dk dynakube.DynaKube) string {
+	var dirName string
+
+	switch {
+	case dk.CodeModulesImage() != "":
+		// An image URI often contains one or several /-s, which is problematic when trying to use it as a folder name.
+		// Easiest to just base64 encode it
+		dirName = base64.StdEncoding.EncodeToString([]byte(dk.CodeModulesImage()))
+	case dk.CodeModulesVersion() != "":
+		dirName = dk.CodeModulesVersion()
+	default:
+		dirName = "unknown"
```

Comment:
will this cause issues in case the same directory name is reused for different runs?

Reply:
What about panicing? - As this should anyway never happen, no need to obscure the error

Reply:
I definitely don't want to panic 😅 
will just return error

Reply:
fixed